### PR TITLE
feat(bProtocol): implement b protocol hodl strategy for LUSD

### PR DESCRIPTION
**Pull request by bergben**
Single asset staking LUSD to fearn fLQTY. 

FYI since strategies get copied around especially by new contributors like me I clean up old strategies too if I encounter irregularities / unnecessary code.

No emergency exit is implemented, the code of bProtocol does not give the possibility to build one. Checked the code though, the deposits should only ever be locked in if bProtocol fails to sufficiently collaterize their trove.

Note the handling of a potential ETH payout to the smart contract because bprotocol eth compounding is not atomic, see comment in code for more.
Liquidation for this is on UniV3: 
image

Run test with `npx hardhat test ./test/bprotocol/bprotocol-lusd-hodl.js` at block `13392555`. 

Test output: 
```
Mainnet BProtocol LUSD HODL
Impersonating...
0xf00dD244228F51547f0563e60bCa65a30FBF5f7f
0xD505FcE5cd73172F748a14FA7113696418362E97
Fetching Underlying at:  0x5f98805A4E8be255a32880FDeC7F6728C6568bA0
New Vault Deployed:  0x5da533Ff9b4b57526B6Be03dBc97BE97ad09854f
reward pool needs to be deployed
New PotPool deployed:  0xc5aC3ad8788C5EB8C8Cc88022C42Ba9a3F51AFaC
Strategy Deployed:  0x82e01223d51Eb87e16A03E24687EDF0F294da6f1
Strategy and vault added to Controller.
    Happy path
-----------------------------
farmerOldBalance 91753.52696417761
farmerOldFLqty 0
in potPool deposited pfLUSD 91753.52696417761
-----------------------------
---------- loop 0 -----------------
old shareprice:  1
new shareprice:  0.9999822871866019
LQTY price: 7.163015137312476
old value:  91753.52696417761
new value:  91753.52696417761
growth:  1
instant APR: 0 %
instant APY: 0 %
fLQTY in potpool:  0
---------- loop 1 -----------------
old shareprice:  0.9999822871866019
new shareprice:  0.9999822871866019
LQTY price: 7.163015137312476
old value:  91753.52696417761
new value:  91772.56316141204
growth:  1.0002074710135322
instant APR: 20.597722223481334 %
instant APY: 22.865383313698693 %
fLQTY in potpool:  2.634008300412196
---------- loop 2 -----------------
old shareprice:  0.9999822871866019
new shareprice:  0.9999822871866019
LQTY price: 7.163015137312476
old value:  91772.56316141204
new value:  91791.58329859207
growth:  1.0002072529798103
instant APR: 20.576075835564556 %
instant APY: 22.838805263584526 %
fLQTY in potpool:  5.26579439667059
---------- loop 3 -----------------
old shareprice:  0.9999822871866019
new shareprice:  0.9999822871866019
LQTY price: 7.163015137312476
old value:  91791.58329859207
new value:  91810.58738926692
growth:  1.0002070352203538
instant APR: 20.554456676726538 %
instant APY: 22.81226636886773 %
fLQTY in potpool:  7.895360163557011
---------- loop 4 -----------------
old shareprice:  0.9999822871866019
new shareprice:  0.9999822871866019
LQTY price: 7.163015137312476
old value:  91810.58738926692
new value:  91829.57544697436
growth:  1.000206817734723
instant APR: 20.532864703296948 %
instant APY: 22.785766553124542 %
fLQTY in potpool:  10.522707474271796
---------- loop 5 -----------------
old shareprice:  0.9999822871866019
new shareprice:  0.9999822871866019
LQTY price: 7.163015137312476
old value:  91829.57544697436
new value:  91848.54748524074
growth:  1.0002066005224792
instant APR: 20.511299871737734 %
instant APY: 22.75930574016318 %
fLQTY in potpool:  13.147838200434993
---------- loop 6 -----------------
old shareprice:  0.9999822871866019
new shareprice:  0.9999822871866019
LQTY price: 7.163015137312476
old value:  91848.54748524074
new value:  91867.50351758099
growth:  1.000206383583184
instant APR: 20.48976213851082 %
instant APY: 22.732883853904106 %
fLQTY in potpool:  15.770754212087457
---------- loop 7 -----------------
old shareprice:  0.9999822871866019
new shareprice:  0.9999822871866019
LQTY price: 7.163015137312476
old value:  91867.50351758099
new value:  91886.44355749869
growth:  1.0002061669164013
instant APR: 20.468251460320648 %
instant APY: 22.706500818598574 %
fLQTY in potpool:  18.39145737769249
---------- loop 8 -----------------
old shareprice:  0.9999822871866019
new shareprice:  0.9999822871866019
LQTY price: 7.163015137312476
old value:  91886.44355749869
new value:  91905.36761848595
growth:  1.0002059505216938
instant APR: 20.44676779376141 %
instant APY: 22.680156558500176 %
fLQTY in potpool:  21.009949564137056
---------- loop 9 -----------------
old shareprice:  0.9999822871866019
new shareprice:  0.9999822871866019
LQTY price: 7.163015137312476
old value:  91905.36761848595
new value:  91924.27571402355
growth:  1.0002057343986273
instant APR: 20.425311095713905 %
instant APY: 22.653850998252167 %
fLQTY in potpool:  23.62623263673306
-----------------------------
-------DONE LOOPING --------
-----------------------------
farmerOldBalance LUSD 91753.52696417761
farmerNewBalance LUSD 91751.90175107589
-----------------------------
Farmer had fLQTY before 0
Farmer has fLQTY after  26.39786391958822
-----------------------------
Farmer total value before 91753.52696417761
Farmer total value after: 91942.68131090568
Farmer total growth:  1.0020615485092135
-----------------------------
Overall APR: 20.467053599471367 %
Overall APY: 22.705031798033804 %
-----------------------------
potpool totalShare:  0
fLQTY in potpool:  7.92358e-13
Farmer earned LUSD:  -1.6252131017312175
Farmer got fLQTY from potpool:  26.39786391958822
-----------------------------
--------- EARNED! -----------
      ✓ Farmer should earn money (36347ms)


  1 passing (43s)
```

### CODE REVIEW

File: contracts/strategies/bprotocol/BProtocolHodlStrategy.sol [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+  function withdrawToVault(uint256 amount) public restricted {
+    // Typically there wouldn't be any amount here
+    // however, it is possible because of the emergencyExit
+    uint256 entireBalance = IERC20(underlying()).balanceOf(address(this));
+
+    if(amount > entireBalance){
+      // While we have the check above, we still using SafeMath below
+      // for the peace of mind (in case something gets changed in between)
+      uint256 needToWithdraw = amount.sub(entireBalance);
+      uint256 toWithdraw = Math.min(rewardPoolBalance(), needToWithdraw);
```
Comment (CryptJS13) [inline contracts/strategies/bprotocol/BProtocolHodlStrategy.sol:151]: Here the ratio between share and LUSD has to be taken into account. Could be done using the `rewardPoolLusdBalance` and `rewardPoolBalance`.

File: contracts/strategies/bprotocol/BProtocolHodlStrategy.sol [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+  */
+  function investedUnderlyingBalance() external view returns (uint256) {
+    if (rewardPool() == address(0)) {
+      return IERC20(underlying()).balanceOf(address(this));
+    }
+    // Adding the amount locked in the reward pool and the amount that is somehow in this contract
+    // both are in the units of "underlying"
+    // The second part is needed because there is the emergency exit mechanism
+    // which would break the assumption that all the funds are always inside of the reward pool
+    return rewardPoolBalance().add(IERC20(underlying()).balanceOf(address(this)));
```
Comment (CryptJS13) [inline contracts/strategies/bprotocol/BProtocolHodlStrategy.sol:110]: Use `rewardPoolLusdBalance` here instead

File: contracts/strategies/bprotocol/BProtocolHodlStrategy.sol [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+import "../../base/interface/IVault.sol";
+import "../../base/upgradability/BaseUpgradeableStrategyUL.sol";
+import "../../base/sushi-base/interfaces/IMasterChef.sol";
+import "../../base/interface/uniswap/IUniswapV2Pair.sol";
+import "../../base/interface/weth/Weth9.sol";
+import "../../base/PotPool.sol";
+import "./interface/IBAMM.sol";
+import "./interface/IStabilityPool.sol";
+
+import "hardhat/console.sol";
```
Comment (CryptJS13) [inline contracts/strategies/bprotocol/BProtocolHodlStrategy.sol:18]: Please remove all the console.logs for prod

File: test/bprotocol/bprotocol-lusd-hodl.js [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+      console.log("Farmer total value before", oldValue.toFixed()/D18.toFixed());
+      console.log("Farmer total value after:", newValue.toFixed()/D18.toFixed());
+      console.log("Farmer total growth: ", newValue.toFixed() / oldValue.toFixed());
+      
+      console.log("-----------------------------");
+
+      Utils.assertBNGt(newValue, oldValue);
+
+      apr = (newValue.toFixed()/oldValue.toFixed()-1)*(24/(blocksPerHour/272))*365;
```
Comment (CryptJS13) [inline test/bprotocol/bprotocol-lusd-hodl.js:214]: Not really important, but for overall apr/apy multiply `blocksPerHour` with `hours`

Comment (bergben) [reply test/bprotocol/bprotocol-lusd-hodl.js]: Thank you. fixed and updated the PR description @CryptJS13

File: contracts/strategies/bprotocol/BProtocolHodlStrategy.sol [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+       * bProtocol fails to correctly collaterize their LQTY trove with ETH
+       */
+  }
+
+  function enterRewardPool() internal {
+    uint256 entireBalance = IERC20(underlying()).balanceOf(address(this));
+    IERC20(underlying()).safeApprove(rewardPool(), 0);
+    IERC20(underlying()).safeApprove(rewardPool(), entireBalance);
+    IBAMM(rewardPool()).deposit(entireBalance);
+    uint bal = IBAMM(rewardPool()).balanceOf(address(this));
```
[the remark this answers is not in the document]
Comment (bergben) [reply contracts/strategies/bprotocol/BProtocolHodlStrategy.sol]: good catch @breadforthepeople, fixed

**Comment by bergben** [comment]
Not merged due to safety concerns regarding ratio / sharePrice manipulation

**Pull request closed without merging**